### DISCUSSION TITLE
Project's inception and its future

### DISCUSSION THREAD

**Discussion by aadcg** [neomacs-project/neomacs#88] [General]
@kchanqvq got in touch some months ago stating that he had developed a structural text editor as a Nyxt extension (see Nyxt's Discord). He requested some architectural changes to Nyxt to accommodate for his extension and I was rather reluctant towards it since he didn't share a demo or the source code to evaluate its quality. I believe https://github.com/atlas-engineer/nyxt/pull/3504 made  him upset. I'm sorry for that again. I reiterated that the starting point to discuss architectural changes to Nyxt must be publishing the extension's source code. 

Eventually, it seems that he gave up on the Nyxt extension idea and started a new project.

The inspiration is clear and some of neomacs's source code files feature the Atlas Engineer copyright header. It's also clear where he has been looking at. For instance, see https://github.com/neomacs-project/neomacs/commit/4c3027df46f4e3b3dcd1526b86c981ba341e75b7 and a commit I have made months earlier https://github.com/atlas-engineer/cl-electron/commit/f62f0bd39f771f44b535abcee471f114d9bcfe16. As of today, `quote-js` became a copy of a [parenscript's method](https://github.com/vsedach/Parenscript/blob/b8dd346d482b730a052902be1af724b16c6f2979/src/printer.lisp#L166).

I think experimentation is key to innovation and it's great to have that neomacs sprouted.

But I'll leave a couple of questions for mediation:

- Is neomacs such a new radical project to exist on its own? Think of Nyxt, Lem, ProjecturEd, Climacs, etc.
- Are we succumbing to the famous "Lisp curse"?

Finally, I'd like to stress that I have come in peace in an attempt to consolidate efforts. Again, I apologize @kchanqvq. Regardless, I wish all the best to neomacs.

**Comment by kchanqvq** [comment]
Thanks for you interest!

> * Is neomacs such a new radical project to exist on its own? Think of Nyxt, Lem, ProjecturEd, Climacs, etc.

I believe the answer is a quite unambiguous yes. It is an experimenting field for lots of new ideas. In fact, maybe Neomacs is radically new enough to qualify as an at least new^3 project. To list a few:

- Neomacs is a structural editor, this makes it unsuitable as an Emacs or Lem extension.

- Neomacs has its own design for modes, which is different from either Lem or Nyxt, and which I quite honestly considered the cleanest reincarnation of Emacs mode so far, and I'm unwilling to give that up. Specifically, Lem has global mode instances and buffer local variables, pretty much reproducing Emacs; Nyxt has a list of buffer local of mode instances; Neomacs uses CLOS mixin.

- The `defstyle` theme facility, which is unlikely to have been developed to the current state if it has been, say, a Nyxt plugin. (https://github.com/atlas-engineer/nyxt/issues/3516)

- Neomacs's goal is to develop a full computing environment, like Lisp machines. This is very different from most other projects you listed. For example, it clearly does not align with Nyxt's priorities.

> * Are we succumbing to the famous "Lisp curse"?

Sure, maybe, but that's maybe also the case for Nyxt which doesn't live as an Emacs plugin on Webkit Xwidget? I found that a lot has been learnt about editor/operating system design in the Neomacs experiments already, and I'm eager to learn more (exchanging structural data between packages, URL-centric operating system like Redox OS...). I don't think these would be possible if Neomacs has been confined to the framework of any existing project.

> Finally, I'd like to stress that I have come in peace in an attempt to consolidate efforts.

Sure! As you can see, I didn't hesitate to build on other project's code and experiences when building Neomacs (I included copyright credits before any such code fragment. I hope that satisfies the original projects' license terms, please point out if I"m mistaken!) You are free to take code and inspirations from us as well :) The project is under currently under GPL because we anticipate we might want to take from GNU Emacs source at some point, but I'm happy to dual license any source files if requested. While the architecture will remain very different, it's quite possible that some extensions can be developed together and I'm open to collaboration!
